network.py

```diff
@@ -0,0 +1,91 @@
+from geopandas import GeoDataFrame
+# street nodes, origin nodes, destination_nodes are the different types of nodes
+
+
+class Network:
+    """
+    A road network with nodes `nodes` and edges `edges` weighted with factor `weight_attribute`,
+    projected onto the `projected_crs` coordinate system.
+
+    Internal class, meant to represent a network within the Zonal object.
+    """
+
+    def __init__(self, nodes: GeoDataFrame, edges: GeoDataFrame, projected_crs: str, weight_attribute=None):
+        self.nodes = nodes
+        self.edges = edges
+        self.crs = projected_crs
+        self.weight_attribute = weight_attribute
+        self.G = None  # G represents a networkx graph. To be deprecated
+        return
+
+    def insert_node(self, label: str, layer_name: str, weight: int):
+        """
+        Adds a node with label `label` to the layer `layer_name` with weight `weight`.
+        Maps to insert_nodes_v2 in madina.py
+        """
+        raise NotImplementedError
+
+    def set_node_value(self, idx, label, new_value):
+        """
+        Sets the node at (`idx`, `label`) value in the network to `new_value`.
+        """
+        self.nodes.at[idx, label] = new_value
+        return
+
+    def create_graph(self, light_graph: bool, dense_graph: bool, d_graph: bool):
+        """
+        TODO: Fill out function spec
+        """
+        raise NotImplementedError
+
+    def get_od_subgraph(self, origin_idx, distance):
+        """
+        Creates a subgraph of the city network `self` from all nodes <= `distance` from the node at `origin_idx`
+
+        Returns:
+            A smaller network graph (?)
+
+        """
+        raise NotImplementedError
+
+    def turn_o_scope(self, origin_idx, search_radius, detour_ratio, turn_penalty=True,
+                     origin_graph=None, return_paths=True):
+        """
+        TODO: Fill out function spec - Unsure what this function does.
```

Comment:
@azizH  was very unsure as to what the function does.

Reply:
Think of it as a modified dijkstra algorithm that include a turn penalty for making a turn along the path (a turn is a function of the visited edges). This is (in my opinion) a network function, that will be used internally by most (or probably all) UNA tools, as it generates the TURN_included_Origin_SCOPE aka: turn_o_scope, what destinations are reachable within the search radius, generate paths and distances to all scope nodes from the origin (pretty much what a dijkstra algorithm do)

Reply:
Ah, that makes sense. I'll move it to the `Network` class - I'm thinking all the network algorithms should be scoped there.